Add Kafka topic encryption proposal (#19)
strimzi/proposals: 015-kafka-topic-encryption.md

## Motivation

Apache Kafka does not directly support any form of encryption-at-rest for data stored at a broker.
Nevertheless, Kafka is increasingly used as a store of data, not just as a
means of transferring it from one location to another. In an
enterprise, this means that Kafka must conform to the same security and
compliance requirements as conventional data storage systems such as relational databases. 
To compensate, Kafka service providers typically use disk or file system encryption. 
This approach has well-known shortcomings, most notably the ability of anyone with appropriate file system permissions, such as system administrators, to read data.  Security best practices and compliance standards
explicitly call for application- or database-level encryption to protect sensitive information against such exposures.
This document proposes a technical solution for providing upper-layer encryption-at-rest for Kafka systems.

## Proposal

An implementation of topic encryption is proposed whereby the message stream between client and broker is
intercepted. Incoming data messages (Produce requests) are inspected to determine whether their payload

> should be encrypted according to a policy.  If so, the data portions are encrypted and the modified

Reviewer: what "data portions" means here? just the kafka message payload? what if sensible information are transferred using Kafka headers in the message?

Author: By "data portions" we are referring to the map of records, keyed by topic, in a Kafka Produce and Fetch messages which are inspected. We have not considered encrypting record headers but this is an interesting suggestion which could be easily accommodated.

As the entire message is navigated to the record level, it would be natural apply encryption to record headers as well. This would require a change to the encryption policy to indicate header names or topic/header combinations to encrypt. The logic change to support header encryption would be minimal.

Author: fyi @SeanRooooney @zrlvep

#### Encrypter-Decrypter
The Encrypter-Decrypter sub-component performs the actual encryption and decryption of messages, implementing a particular encryption algorithm and making use of the KMS service to obtain topic keys.

#### Key Service
The Key Service is used in the Encryption Module to handle interactions with a key management system and to perform optional primitives such as those related to envelope encryption. The interface is generic in order to accommodate various key management implementations.

#### Policy Service
Encrypter-decrypters require a means to lookup encryption requirements and parameters for a topic. As a corresponding standard does not exist, the core component will interact with a generic internal interface for retrieving policy on how a topic is, or not is, to be encrypted. Implementers can develop and configure their own implementation of the policy service, be it a remote microservice, a local pre-configured cache, a cluster registry, etc.

For each topic to be encrypted, a policy exists detailing:

> - topic name

Reviewer 2: I can see use cases where it would be convenient to have a single policy governing multiple topics, e.g. all topics beginning with a given prefix should use the same policy.

Author: I Kafka record is comprised of Key and Value fields. Encryption is applied to the Value field only. The message key therefore is the same across encryptions.  

As an example, the Kafka javadoc for ProducerRecord defines Key and Value as distinct fields:
https://kafka.apache.org/10/javadoc/org/apache/kafka/clients/producer/ProducerRecord.html

Author: We have considered supporting a regular expression for topic names. Because regular expressions can result in multiple matches and thus can be a bit unpredictable, we initially have taken a conservative stance that either policy is explicitly specified for each topic or a global wildcard (*) policy is applied.  We are definitely open to discussion. We have also considered the idea of a default policy.  Supporting multiple topics does seem to make sense either as a comma-separated list of topic names or the slightly risky regex.

Author: We leave open the full functionality of the policy service. The matching expression does not have to be an exact match or a regular expression but rather also some sort of query expression. Same caveat with regular expressions apply as mentioned above apply - a topic may match more than one expression and therefore more than one policy.  The convenience of regular and query expressions must be weighed with potential ambiguities which may arise.

Reviewer 3: I agree that things like regexes or prefixes would need to be handled carefully since they open a big space for misconfiguration.

<!-- This template is provided as an example with sections you may wish to comment on with respect to your proposal. Add or remove sections as required to best articulate the proposal. -->

# Proxy-Based Kafka Per-Topic Encryption

The goal of this proposal is to provide topic-level encryption-at-rest for Kafka such that distinct encryption keys can be used for different topics. 
A core library is proposed which is deployed into a proxy. Proxies can flexibly be deployed in a variety of topologies and require no changes to Kafka clients and brokers. All encryption/decryption takes place at the proxy. An unencrypted topic's messages pass through the proxy without being changed and without

> paying any performance penalty.

Author: Agree. I will include more precise statement in the next update.

used in testing, is shown below:

```
[
 {
   "topic" : "juicy-topic",
   "kms_url" : "https://kms.abc.com",
   "kms_instance": "11111111-2222-3333-4444-55555555555",
   "key_ref" : "e2a73a6c4bf9",
   "credential" : "dkT4WSCnrlNE7"
 },
 {
   "topic" : "creditcard-data",
   "kms_url" : "https://kms.abc.com",
   "kms_instance": "11111111-2222-3333-4444-55555555555",
   "key_ref" : "9957cdf5f32a",
   "credential" : "dkT4WSCnrlNE7"
 }
]
```

In this example, two topics will be encrypted with two different keys residing in the same KMS.

### Message Handling
Message handling is concerned with facilities for intercepting messages, deserializing and inspecting them, invoking encryption as accorded by policy, modifying messages, and finally serializing and forwarding messages onward to Kafka brokers and clients. To accomplish encryption and decryption of data payloads, only Kafka *Produce requests* and *Fetch responses* need to be examined and potentially modified. High quality open source libraries exist already for implementing message introspection and serialization/deseriaization, e.g.,
https://github.com/Shopify/sarama.

> ### Message Encryption

Reviewer 3: Maybe a stupid question, but I do not have much experience with encryption other then TLS connections. Is there any impact on this from things like:
* Expiration of the encryption keys?
* Keys needing to be replaced because they were leaked / compromised?

Or is this not something we need to address here?

Reviewer 4: Partially. The idea is to let implementations of the key service interface deal with such requirements. I.e. implement envelope encryption to be able to expire root keys or support multiple different keys per topic.

I think one thing we need to mention is, that re-encryption of topics is not supported by this proposal. Meaning, if a key is compromised, data encrypted with it is compromised too and can only be mitigated by copying the topic and using a different encryption key.

Reviewer 3: Yeah, I think it would be great to mention this to make it more clear.

As described in [Kafka message format documentation](https://kafka.apache.org/documentation/#messageformat), data is transmitted in [record batches](https://kafka.apache.org/documentation/#recordbatch). Within a batch, each [record](https://kafka.apache.org/documentation/#record) is comprised of several fields of which two may contain data and therefore are of direct relevance for topic encryption: the _value_ field and the _headers_ field, an array of optional [record header structures](https://kafka.apache.org/documentation/#recordheader). _Key_ fields are never encrypted in this proposal. Any functionality relying on keys such as partitioning and compaction are unaffected and thus preserved by topic encryption.

An example is now described to illuminate the essential details of the message encryption process.

> The encryption module is embedded in a proxy and the proxy, as an intermediary, terminates  connections between clients and the broker, placing the proxy in the position of intercepting all Kafka traffic. The message handling component examines all messages to identify the apikey (i.e., the Kafka message type). All message types besides Produce and Fetch are  immediately forwarded without any processing. With the help of the message handling library, Produce requests are deserialized and inspected more deeply to identify the topic for which records are destined.

Reviewer 2: > All message types besides Produce and Fetch are immediately forwarded without any processing.

You probably want to do something with `ApiVersions` requests. They're used to inform the client about the protocol versions (for each api key) supported by the broker. So it would be possible for a new broker version to add a new version of a `Produce` request which the proxy didn't understand, and by just sending the `ApiVersions` response to the client unchanged a client which also supported that version of `Produce` would use it. So for the `Produce` and `Fetch` APIs the max version in the `ApiVersions` response should the `min(brokerMaxVersion, proxyMaxVersion)`, likewise the min versions should be the `max(brokerMinVersion, proxyMinVersion)`. If you're using Envoy it's possible that it already handles this, but it's worth checking.

Author: Agree, good point. Will add to the next update.

If a topic matches a topic in the encryption policy, the contents of the _value_ field of each record is encrypted and replaced with the corresponding ciphertext. Metadata about the encryption (e.g., encrypted key or key reference, nonce) is added to the record's headers in order that this information is available later for decryption.

The encryption metadata is currently stored in Kafka record headers. Record headers are optional name value/pairs stored in an array associated with the record. Record header names are not required to be unique, thus name "collisions" are possible. Ordering and versioning are used to avoid problems arising from names used for metadata which coincidentally equal names chosen by a Kafka client. Ordering means that the record header is rewritten such that the encryption-related headers
occur first in the array. The first header always contains a version from which the exact
metadata field names can be derived. The decryption algorithm therefore can discern between
headers added by the encryption module and those added by the client. The version field

> is enables future migration and backwards compatibility should changes in semantics occur such as modified message formats or encryption options.

Reviewer 4: Yes, added headers will be removed before forwarding them to the consumers.

> is enables future migration and backwards compatibility should changes in semantics occur such as modified message formats or encryption options.

Reviewer 3: Ok, great. Maybe you could add it to the proposal to make it more clear.

Encrypting header names as well would likewise be a relatively simple extension however we recommend further investigating the full implications of encrypting header names.

### Embedding Topic Encryption in a Proxy
A proxy can be deployed as a free-standing process or as a sidecar in a Kubernetes pod. 
In both cases, the proxy needs to intercept and de-serialize Kafka messages, encrypt/decrypt the relevant records, 
and reserialize them into a new message. The overhead of a proxy potentially has an effect on performance.
It also means that the proxy version must always be in phase with that of the broker.
A proxy-based solution however has the advantage that both Kafka client and broker are unaware
of the proxy and do not require any modification or configuration to support encryption at rest.

Whether the proxy runs under the control of the client or the broker is considered a configuration

> issue, i.e. exactly the same proxy would be used in both cases, but when under the control of the

Reviewer 3: I think there are some notable difference depending on where it is running which should be covered here:

* When running as a sidecar inside the broker pod, the proxy has to listen to a single port which will only get the connections for this particular broker. This can be also easily orchestrated and this port / address might be used as a advertised host in the broker configuration. And thus the proxy really needs to only deal with the _Produce_ and _Fetch_ requests.

* But on the client side, the advertised hosts metadata make it a way more complicated. They would essentially lead the clients to bypass the proxy and connect to the brokers directly to produce / fetch messages. So I think this would need some special handling, or? I guess there will be two options. One the proxy can just open more different ports (one per broker) and modify the advertised hosts in metadata. Or I guess it can approach it Istio-style and just high-jack all connections into the proxy. But that would need to be able to filter out non-Kafka connections and possibly also connections to other Kafka clusters which should not be encrypted.

Or did I missed some option?

Author: @scholzj No, you did not miss an option and yes this requires special handling. Kafka clients need a broker list and broker metadata which connects them through one or more proxies. Meanwhile, the brokers  need to interact with each other and answer metadata requests to their clients. This can be approached from different angles. We have viewed this as an integration issue and outside the scope of the core encryption proposal. Although we have deployed an implementation, I am unsure whether we are at liberty to provide details. I will clarify. In any case, this is an area where we very much welcome the expertise of the strimzi community. In the meantime, we will add a section to the proposal to call this out and possibly outline potential solutions.

Reviewer 3: Sounds like a plan. I'm fine if there are not too many details ... but as you said, might be just worth mention either that it is out of scope or just outline the possible solutions and leave the details for implementation.

> client it would be their responsibility to ensure that producers/consumers are passing through a proxy with the same configuration.

Reviewer 2: Client identity might be a problem here. When the proxy is just proxying a single client it can forward the `SaslHandshake` and `SaslAuthenticate` messages, so SASL authentication (configure on the client) should just work. Likewise TLS authentication would work fine, because the proxy can be configured with the TLS keystore and truststore which the client would have used. But it gets more complicated when more than a single client it using the same proxy (and that could even happen in the sidecar deployment, if there were multiple client instances being used in the main container of the pod). You probably don't want to worry about this now, which is fine, but I think we should be explicit about that.

Similarly there can be issues with client quotas if the proxy is serving multiple clients but using a single identity for its connection to the broker.

Author: You raise important integration issues which will have to be addressed. Indeed, they may vary depending on how the proxy is deployed.

means of transferring it from one location to another. In an
enterprise, this means that Kafka must conform to the same security and
compliance requirements as conventional data storage systems such as relational databases. 
To compensate, Kafka service providers typically use disk or file system encryption. 
This approach has well-known shortcomings, most notably the ability of anyone with appropriate file system permissions, such as system administrators, to read data. Security best practices and compliance standards
explicitly call for application- or database-level encryption to protect sensitive information against such exposures.
This document proposes a technical solution for providing upper-layer encryption-at-rest for Kafka systems.

## Proposal

### Overview
An implementation of topic encryption is proposed whereby the message stream between client and broker is intercepted. Incoming data messages (Produce requests) are inspected to determine whether their payload
should be encrypted according to a policy. If so, the data portions are encrypted and the modified
message is forwarded to the broker. As a result, the topic data is stored by the broker in encrypted form.

> On the reverse direction, encrypted responses (responses to Fetch requests) are decrypted prior to being sent to clients.

Author: @tombentley These are all important points you raise. 

- Agree, policy must be dynamically updated to, at the very least, accommodate topic creation and deletion. This should be done without restarting.  
- The proposal includes the KMS interface and at least one implementation of that interface initially, IBM Key Protect. We have started work with IBM Secrets Manager which is Vault. Very glad to discuss whether Vault would be a preferred initial KMS implementation.
- Migrating a key to another KMS can be supported however could require a restart, at least initially. The new KMS must of course be of a supported type. As long as the key material remains the same, migration is a configuration change.
- Yes, the credential is very sensitive data. It is used to access the KMS. Therefore the entire policy, in its current form,  must be treated as a secret.
- kms_instance is an optional parameter. To access IBM Key Protect, a URL and an instance ID are required. If a KMS does not require this parameter (e.g., Vault) it can be dropped. Perhaps there are other KMS systems out there which require an additional parameter in addition to a URL. kms_instance is a placeholder for such a param. 

I will provide an update with a brief description of the policy parameters.

> On the reverse direction, encrypted responses (responses to Fetch requests) are decrypted prior to being sent to clients.

Reviewer 2: Thanks!

I think there's been most user interest (i.e. for TLS-related stuff) around Vault (@scholzj correct me if I'm wrong), so it would be great to have support for that.

I guess the KMS migration via restart should be fine (since we don't expect it to be a common thing). I hadn't realised that the way to do this would be to just export the old keys etc from the one to the other.

> On the reverse direction, encrypted responses (responses to Fetch requests) are decrypted prior to being sent to clients.

Author: KMS semantics should be the same as well when migrating. If AES data encryption keys are stored in the KMS, then migrating should be straightforward. However if an encryption model is employed which involves some sort of computation on the part of the KMS (e.g., envelope encryption with wrap/unwrap primitives) then one KMS is  presumably not a drop-in replacement since vendors may implement functions (e.g., envelope encryption) differently. The topic would then contain cryptographic artifacts from both KMS systems and could not be supported.  So migration needs to be qualified.

> On the reverse direction, encrypted responses (responses to Fetch requests) are decrypted prior to being sent to clients.

Author: @tombentley fyi I have updated the proposal to include clarification on policy updates.

> On the reverse direction, encrypted responses (responses to Fetch requests) are decrypted prior to being sent to clients.

Reviewer 2: @chris-giblin could you fix the typo?

explicitly call for application- or database-level encryption to protect sensitive information against such exposures.
This document proposes a technical solution for providing upper-layer encryption-at-rest for Kafka systems.

## Proposal

### Overview
An implementation of topic encryption is proposed whereby the message stream between client and broker is intercepted. Incoming data messages (Produce requests) are inspected to determine whether their payload
should be encrypted according to a policy. If so, the data portions are encrypted and the modified
message is forwarded to the broker. As a result, the topic data is stored by the broker in encrypted form.
On the reverse direction, encrypted responses (responses to Fetch requests) are decrypted prior to being sent to clients.

As defined in the encryption policy, each topic can be encrypted by a different key,
allowing brokers to store a mix of encrypted and unencrypted data, where
data owners can manage the keys to their topics.

> Keys ideally are stored in a key management system with access policies and logging. 

Reviewer 2: ```suggestion
Keys will be stored in an external key management system with access policies and logging. 
```

![overview](images/015-kafkaenc-overview.png)

### Components
#### Encryption Module
The Encryption Module is the top-level component which encapsulates encryption functionality and is embedded in a proxy.
A central design goal is adaptability through modularity. Encrypter/decrypter(s), the policy metadata service and the key management service (KMS) are expressed as interfaces and pluggable. The Encryption Module instantiates these components during initialization, allowing a configurable combination of KMS, metadata and encryption implementations.

#### Encrypter-Decrypter
The Encrypter-Decrypter sub-component performs the actual encryption and decryption of messages, implementing a particular encryption algorithm and making use of the KMS service to obtain topic keys.

#### Key Service
The Key Service is used in the Encryption Module to handle interactions with a key management system and to perform optional primitives such as those related to envelope encryption. The interface is generic in order to accommodate various key management implementations.

#### Policy Service

> Encrypter-decrypters require a means to lookup encryption requirements and parameters for a topic. As a corresponding standard does not exist, the core component will interact with a generic internal interface for retrieving policy on how a topic is, or not is, to be encrypted. Implementers can develop and configure their own implementation of the policy service, be it a remote microservice, a local pre-configured cache, a cluster registry, etc.

Reviewer 2: ```suggestion
Encrypter-decrypters require a means to lookup encryption requirements and parameters for a topic. As a corresponding standard does not exist, the core component will interact with a generic internal interface for retrieving policy on how a topic is, or is not, to be encrypted. Implementers can develop and configure their own implementation of the policy service, be it a remote microservice, a local pre-configured cache, a cluster registry, etc.
```
